### PR TITLE
feat: add new `Target.metadata` attribute

### PR DESCRIPTION
**Pull request by hiltonc**
This adds target metadata tags. It is needed for https://github.com/tuist/tuist/pull/6680.

### CODE REVIEW

**Review by danieleformichelli — APPROVED** [review]

File: Sources/XcodeGraph/Models/Metadata/TargetMetadata.swift [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/// The metadata associated with a target.
+public struct TargetMetadata: Codable, Equatable {
+    public var tags: Set<String>
+
+    public init(
```
Comment (pepicrft) [inline Sources/XcodeGraph/Models/Metadata/TargetMetadata.swift:5]: To be consistent with the other models, I'd make this one internal and add a static initializer:

```swift
let target = Target(metadata: .metadata(tags: ["foo"]))
```

Considering we'll use this interface to codify things like "ownership" of a target, for example by specifying a team, I wonder if we should have tags instead, some of which have special semantics, or add those attributes to `TargetMetadata` instead. For example:

```swift
let target = Target(metadata: .metadata(tags: ["foo", "bar"], team: "my-team"))
let target = Target(tags: ["foo", "bar", .team("my-team")]
```

I kind of like the latter more. Thoughts @tuist/core ?

**Review by pepicrft — CHANGES_REQUESTED** [review]
I brought up some thoughts about the API to get everyone's thoughts. Thanks for taking the lead on executing this one @hiltonc 👏🏼

File: Sources/XcodeGraph/Models/Metadata/TargetMetadata.swift (hunk 1 above)
Comment (kwridan) [reply Sources/XcodeGraph/Models/Metadata/TargetMetadata.swift:5]: Looking at the original issue it seems the motivation for this addition was to allow grouping targets when generating focused projects as opposed to codifying team ownership specifically. I wonder if a different solution that takes into account schemes (which also has the concept of grouping related targets) can be used there 🤔

If we're proceeding with tags / metadata anyways - I would personally keep it general and not too specific with encoding specific concepts like "team" which may not be applicable in the generation context as this won't scale (e.g. if another team / feature needs a different semantic tag they need to update Tuist Graph / Core).

Comment (danieleformichelli) [reply Sources/XcodeGraph/Models/Metadata/TargetMetadata.swift:5]: I tend to agree that schemas might be the right answer to this problem, as we can already focus on a specific schema.
Is there anything that is not covered by that?

Comment (pepicrft) [reply Sources/XcodeGraph/Models/Metadata/TargetMetadata.swift:5]: The motivation is to introduce a layer of semantics that features like project generation can leverage to filter targets or graph to visually group elements in the graph. Then, Tuist Server will leverage this to introduce semantics, for example, the team that owns a particular module. We can define custom conventions for those, which developers would need to follow in their tag strings (e.g. `team:1234`), but that's prone to typos. They'd map to the string convention internally.

What if we define those as extensions over the base type? (i.e., like we do with settings)

**Review by pepicrft — APPROVED** [review]

**Pull request merged**